x-pack/plugins/endpoint/server/routes/resolver/utils/fetch.ts

```diff
@@ -5,117 +5,176 @@
     }
 
     const query = new LifecycleQuery(this.indexPattern, this.endpointID);
-    const { results } = await query.search(this.client, curNode);
+    const results = await query.search(this.client, curNodeID);
 
     if (results.length === 0) {
-      tree.setNextAncestor(null);
       return;
     }
-    tree.addAncestor(previousNode, ...results);
+    ancestorInfo.ancestors.push(createLifecycle(curNodeID, results));
 
     const next = parentEntityId(results[0]);
-    if (next !== undefined) {
-      await this.doAncestors(tree, next, curNode, levels - 1);
+    if (next === undefined) {
+      return;
     }
+    await this.doAncestors(next, levels - 1, ancestorInfo);
   }
 
-  private async doEvents(tree: Tree, limit: number, after?: string) {
+  private async doEvents(limit: number, after?: string) {
     const query = new EventsQuery(
+      PaginationBuilder.createBuilder(limit, after),
       this.indexPattern,
-      this.endpointID,
-      getPaginationParams(limit, after)
+      this.endpointID
     );
 
     const { totals, results } = await query.search(this.client, this.id);
-    tree.addEvent(...results);
-    tree.paginateEvents(totals, results);
-    if (results.length === 0) tree.setNextEvent(null);
+    if (results.length === 0) {
+      // return an empty set of results
+      return createRelatedEvents(this.id);
+    }
+    if (!totals[this.id]) {
+      throw new Error(`Could not find the totals for related events entity_id: ${this.id}`);
+    }
+
+    return createRelatedEvents(
+      this.id,
+      results,
+      PaginationBuilder.buildCursor(totals[this.id], results)
+    );
   }
 
   private async doChildren(
-    tree: Tree,
+    cache: ChildrenNodesHelper,
     ids: string[],
     limit: number,
     levels: number,
     after?: string
   ) {
-    if (levels === 0 || ids.length === 0) return;
+    if (levels === 0 || ids.length === 0) {
+      return;
+    }
 
     const childrenQuery = new ChildrenQuery(
+      PaginationBuilder.createBuilder(limit, after),
       this.indexPattern,
-      this.endpointID,
-      getPaginationParams(limit, after)
+      this.endpointID
     );
     const lifecycleQuery = new LifecycleQuery(this.indexPattern, this.endpointID);
 
-    const { totals, results } = await childrenQuery.search(this.client, ...ids);
+    const { totals, results } = await childrenQuery.search(this.client, ids);
     if (results.length === 0) {
-      tree.markLeafNode(...ids);
       return;
     }
 
     const childIDs = results.map(entityId);
-    const children = (await lifecycleQuery.search(this.client, ...childIDs)).results;
+    const children = await lifecycleQuery.search(this.client, childIDs);
 
-    tree.addChild(...children);
-    tree.paginateChildren(totals, results);
-    tree.markLeafNode(...childIDs);
+    cache.addChildren(totals, children);
 
-    await this.doChildren(tree, childIDs, limit * limit, levels - 1);
+    await this.doChildren(cache, childIDs, limit, levels - 1);
```

Comment:
Instead of doing `limit * limit` here, it's only allowing a static number of children nodes in the response at each level. This is temporary because as soon as `msearch` and the `ancestry` array is implemented in the next PR this function is going to completely change anyway.